### PR TITLE
Scripts/UBRS: The Beast

### PR DESCRIPTION
**Pull request by Riztazz**
**Target branch(es):** 3.3.5/master

- [X] 3.3.5

**Issues addressed:** Closes #  (insert issue tracker number)
#19387

**Tests performed:** (Does it build, tested in-game, etc.)
Builds and works in game

**Known issues and TODO list:** (add/remove lines as needed)
Yeah this entire instance code is just clusterfuck

### CODE REVIEW

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp [hunk 1]
```diff
@@ -19,21 +19,62 @@
 #include "ScriptMgr.h"
 #include "blackrock_spire.h"
 #include "ScriptedCreature.h"
+#include "Player.h"
```
Comment (Riztazz) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp]: Yeah i need to clean this up, my msvc was acting up on dynamic

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp [hunk 2]
```diff
@@ -48,17 +89,88 @@ class boss_the_beast : public CreatureScript
...
+
+                        if (nearbyOrcsGUIDs.empty())
+                            FindNearbyOrcs();
+
+                        for (ObjectGuid guid : nearbyOrcsGUIDs)
+                        {
+                            if (Creature* orc = ObjectAccessor::GetCreature(*me, guid))
+                            {
+                                orc->GetMotionMaster()->MovePoint(1, orc->GetRandomPoint(OrcsRunawayPosition, 5.0f));
+                                orc->m_Events.AddEvent(new OrcDeathEvent(orc), me->m_Events.CalculateTime(6 * IN_MILLISECONDS));
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp]: has no support of std::chrono here?

Comment (Riztazz) [reply src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp]: Nope

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp [hunk 3]
```diff
@@ -100,10 +220,69 @@ class boss_the_beast : public CreatureScript
...
+        {
+            std::vector<Creature*> temp;
+            me->GetCreatureListWithEntryInGrid(temp, NPC_BLACKHAND_ELITE, 50.0f);
+
+            for (Creature* creature : temp)
+                nearbyOrcsGUIDs.push_back(creature->GetGUID());
+        }
+
+    private:
+        bool beastReached;
```
Comment (sirikfoll) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp]: private members should have the _ prefix. Nice job btw 😄

Comment (sirikfoll) [reply src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp]: I think the new script register model needs to be used for all the new scripts at least. If you have the time, you could also convert those you are updating haha

Comment (Riztazz) [reply src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp]: aight

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp [hunk 4]
```diff
@@ -20,90 +20,257 @@
...
+        _EnterCombat();
+        events.ScheduleEvent(EVENT_FLAME_BREAK, Seconds(12));
+        events.ScheduleEvent(EVENT_IMMOLATE, Seconds(3));
+        events.ScheduleEvent(EVENT_TERRIFYING_ROAR, Seconds(23));
+        events.ScheduleEvent(EVENT_BERSERKER_CHARGE, Seconds(2));
+        events.ScheduleEvent(EVENT_FIREBALL, Seconds(8), Seconds(21));
+        events.ScheduleEvent(EVENT_FIREBLAST, Seconds(5), Seconds(8));
+    }
+
+    void JustDied(Unit* /*killer*/) override
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp]: not needed

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp [hunk 5]
```diff
@@ -20,90 +20,257 @@
...
+            switch (eventId)
+            {
+                case EVENT_FLAME_BREAK:
+                    DoCastVictim(SPELL_FLAMEBREAK);
+                    events.ScheduleEvent(EVENT_FLAME_BREAK, Seconds(10));
+                    break;
+                case EVENT_IMMOLATE:
+                    if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 100.f, true))
+                        DoCast(target, SPELL_IMMOLATE);
+                    events.ScheduleEvent(EVENT_IMMOLATE, Seconds(8));
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp]: events.Repeat(Seconds(8));

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp [hunk 6]
```diff
@@ -20,90 +20,257 @@
...
+            return;
+
+        while (uint32 eventId = events.ExecuteEvent())
         {
-            _JustDied();
+            switch (eventId)
+            {
+                case EVENT_FLAME_BREAK:
+                    DoCastVictim(SPELL_FLAMEBREAK);
+                    events.ScheduleEvent(EVENT_FLAME_BREAK, Seconds(10));
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp]: events.Repeat(Seconds(10));

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp [hunk 7]
```diff
@@ -20,90 +20,257 @@
...
+                    break;
+                case EVENT_IMMOLATE:
+                    if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 0, 100.f, true))
+                        DoCast(target, SPELL_IMMOLATE);
+                    events.ScheduleEvent(EVENT_IMMOLATE, Seconds(8));
+                    break;
+                case EVENT_TERRIFYING_ROAR:
+                    DoCastVictim(SPELL_TERRIFYINGROAR);
+                    events.ScheduleEvent(EVENT_TERRIFYING_ROAR, Seconds(20));
```
Comment (Keader) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp]: events.Repeat(Seconds(20));

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/instance_blackrock_spire.cpp [hunk 8]
```diff
@@ -283,6 +287,11 @@ class instance_blackrock_spire : public InstanceMapScript
                         if (GetBossState(DATA_DRAGONSPIRE_ROOM) != DONE)
                             Events.ScheduleEvent(EVENT_DARGONSPIRE_ROOM_STORE, 1000);
                     }
+                    break;
+                case DATA_SPAWN_FINKLE_EINHORN:
+                    if (TempSummon* finkle = instance->SummonCreature(NPC_FINKLE_EINHORN, finkleSpawnPosition))
```
Comment (Killyana) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/instance_blackrock_spire.cpp]: https://woehead.way-of-elendil.fr/?spell=16710

Comment (tkrokli) [reply src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/instance_blackrock_spire.cpp]: (http://www.wowhead.com/spell=16710/finkle-is-einhorn)

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/instance_blackrock_spire.cpp [hunk 9]
```diff
@@ -28,6 +28,8 @@
 
 uint32 const DragonspireMobs[3] = { NPC_BLACKHAND_DREADWEAVER, NPC_BLACKHAND_SUMMONER, NPC_BLACKHAND_VETERAN };
 
+const Position finkleSpawnPosition = { 58.259682f, -535.346008f, 110.935753f, 3.247705f };
+
```
Comment (offl) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/instance_blackrock_spire.cpp]: Unused after e6878a60474bbb34f6bf2e7fd81aaf259c6ca341. Also `bool finkleSpawned` is unused too.

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp [hunk 10]
```diff
@@ -20,90 +20,253 @@
...
     }
 
-    struct boss_thebeastAI : public BossAI
+private:
+    Creature* _me;
+};
+
+struct boss_the_beast : public BossAI
+{
+    boss_the_beast(Creature* creature) : BossAI(creature, DATA_THE_BEAST) { }
```
Comment (jackpoz) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/boss_the_beast.cpp]: please initialize the class member fields in the constructor

File: src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/instance_blackrock_spire.cpp [hunk 11]
```diff
@@ -510,6 +518,8 @@ class instance_blackrock_spire : public InstanceMapScript
             ObjectGuid runecreaturelist[7][5];
             ObjectGuid go_portcullis_active;
             ObjectGuid go_portcullis_tobossrooms;
+
+            bool finkleSpawned;
```
Comment (jackpoz) [inline src/server/scripts/EasternKingdoms/BlackrockMountain/BlackrockSpire/instance_blackrock_spire.cpp]: please delete this field is it's no longer used

**Review by jackpoz — CHANGES_REQUESTED** [review]
There are a couple of things to be addressed in this PR, other than that it looks fine.
Is it ready to be merged (tested/no more WIP) ?

**Comment by Riztazz** [comment]
it is ready for merge

**Comment by tkrokli** [comment]
The AppVeyor "error" happened because of a 60 minute time-out, not because of any content in this PR. (It is likely that AppVeyor can finish normally if a new change is committed, like renaming the SQL file.)

**Comment by jackpoz** [comment]
> It is likely that AppVeyor can finish normally if a new change is committed

appveyor builds can be manually restarted

**Comment by tkrokli** [comment]
Very well, thanks for the info. I was looking around while logged in to my own AppVeyor account, but I didn't find a restart option like the one in travis-ci. I need to read more documentation, I suppose.

**Comment by jackpoz** [comment]
DDuarte has to grant you permissions to log in as his account to be able to do that

**Pull request merged**